src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/Telemetry/VsTelemetryService.cs

```diff
@@ -1,20 +1,142 @@
+using System.Security.Cryptography;
+using System.Text;
+using System.Threading.Tasks;
+using Microsoft.VisualStudio.ProjectSystem;
 
 namespace Microsoft.VisualStudio.Telemetry
 {
     [Export(typeof(ITelemetryService))]
-    internal class VsTelemetryService : ITelemetryService
+    [Export(typeof(IVsTelemetryService))]
+    internal class VsTelemetryService : IVsTelemetryService, ITelemetryService
     {
+        private const string EventPrefix = "vs/projectsystem/managed/";
+        private const string PropertyPrefix = "VS.ProjectSystem.Managed.";
+        private readonly IUnconfiguredProjectCommonServices _unconfiguredProjectCommonServices;
+        private readonly object _lock =  new object();
+
+        private TelemetryEventCorrelation _telemetryEventCorrelation;
+        private bool _telemetryEventCorrelationInitialized;
+        private Guid _projectGuid;
+
+        [ImportingConstructor]
+        public VsTelemetryService(IUnconfiguredProjectCommonServices unconfiguredProjectCommonServices)
+        {
+            Requires.NotNull(unconfiguredProjectCommonServices, nameof(unconfiguredProjectCommonServices));
+
+            _unconfiguredProjectCommonServices = unconfiguredProjectCommonServices;
+        }
+
+        /// <summary>
+        /// Creates the Project Correlation asset which will be used to correlate all the telemetry events for this project
+        /// identified uniquely through the Project Guid.
+        /// </summary>
+        /// <returns></returns>
+        private async Task<bool> CreateProjectCorrelationAssetAsync()
+        {
+            var configurationGeneralProperties = await _unconfiguredProjectCommonServices
+                                                .ActiveConfiguredProjectProperties
+                                                .GetConfigurationGeneralPropertiesAsync()
+                                                .ConfigureAwait(false);
+
+            if (Guid.TryParse((string)await configurationGeneralProperties.ProjectGuid.GetValueAsync().ConfigureAwait(false), out Guid guid))
+            {
+                _projectGuid = guid;
+            }
+            else
+            {
+                return false;
+            }
+
+            var targetFrameworkProperty = await configurationGeneralProperties.TargetFramework.GetEvaluatedValueAtEndAsync().ConfigureAwait(false);
+            var targetFrameworksProperty = await configurationGeneralProperties.TargetFrameworks.GetEvaluatedValueAtEndAsync().ConfigureAwait(false);
+            var outputTypeProperty = await configurationGeneralProperties.OutputType.GetEvaluatedValueAtEndAsync().ConfigureAwait(false);
+
+            lock (_lock)
+            {
+                if (!_telemetryEventCorrelationInitialized)
+                {
+                    var eventName = "Project";
+                    var projectProperties = new Dictionary<string, object>();
+
+                    if (!string.IsNullOrEmpty(targetFrameworksProperty))
```

Comment:
Thats good question. The asset has a property called AssetID, which is the ProjectGuid. This asset also has a SessionId which will be unique across different instances of the Project instantiation. There are other Ids, which I have not looked into but I dont think the properties of the asset will be part of <asset identity>, but I am not sure. But the properties themselves are useful because it helps us understand the initial state of the project.